### PR TITLE
cephadm: add start/stop hooks and c-v activate on container start

### PR DESCRIPTION
**Pull request by liewegas**
- should allow dm-crypt to work
- fix pod stop
- fix behavior with no /etc/ceph/ceph.conf

### CODE REVIEW

File: src/ceph-daemon/ceph-daemon [hunk 1]
```diff
@@ -897,14 +918,18 @@ class CephContainer:
                  args=[],
                  volume_mounts={},
                  cname='',
-                 container_args=[]):
-        # type: (str, str, List[str], Dict[str, str], str, List[str]) -> None
+                 container_args=[],
+                 prestart=None,
+                 poststop=None):
+        # type: (str, str, List[str], Dict[str, str], str, List[str], Optional[CephContainer], Optional[CephContainer]) -> None
```
Comment (sebastian-philipp) [inline src/ceph-daemon/ceph-daemon]: :-1: here. `prestart` and `poststop` should not be part of `CephContainer`:

* Now we're leaving the realm of "Containers" and head into the realm of "Pods" here. There is a reason K8s differentiates them
* `prestart` and `poststop` are completely unused the rest of the methods
* `CephContainer` is now a tree data structure: 
    ```python
    CephContainer(
        prestart=CephContainer(
            poststop=CephContainer()
        ), 
        poststop=CephContainer(
            prestart=CephContainer()
        )
    )
    ```
    This looks silly. 

Instead, what about creating a `CephPod` class:

```python
class CephPod(object):
    def __init__(
            self,
            unit_name,  # type: str
            data_dir,  # type: str
            app,  # type: CephContainer
            prestart=None,  # type: Optional[CephContainer]
            poststop=None,  # type: Optional[CephContainer]
    ):
        self.unit_name = unit_name
        self.data_dir = data_dir
        self.app = app
        self.prestart = prestart
        self.poststop = poststop

    def deploy_daemon_units(self, fsid, uid, gid,
                            enable=True, start=True):
        # type: (str, int, int, bool, bool) -> None

        # cmd
        with open(self.data_dir + '/unit.run', 'w') as f:
            if self.prestart:
                f.write(' '.join(self.prestart.run_cmd()) + '\n')
            f.write(' '.join(self.app.run_cmd()) + '\n')
            os.fchmod(f.fileno(), 0o600)
        with open(self.data_dir + '/unit.poststop', 'w') as f:
            if self.poststop:
                f.write(' '.join(self.poststop.run_cmd()) + '\n')
            os.fchmod(f.fileno(), 0o600)

        # systemd
        install_base_units(fsid)
        unit = get_unit_file(fsid, uid, gid)
        unit_file = 'ceph-%s@.service' % (fsid)
        with open(args.unit_dir + '/' + unit_file + '.new', 'w') as f:
            f.write(unit)
            os.rename(args.unit_dir + '/' + unit_file + '.new',
                      args.unit_dir + '/' + unit_file)
        call_throws(['systemctl', 'daemon-reload'])

        call(['systemctl', 'stop', self.unit_name],
             verbose_on_failure=False)
        call(['systemctl', 'reset-failed', self.unit_name],
             verbose_on_failure=False)
        if enable:
            call_throws(['systemctl', 'enable', self.unit_name])
        if start:
            call_throws(['systemctl', 'start', self.unit_name])
```

Comment (jan--f) [reply src/ceph-daemon/ceph-daemon]: Hmm I disagree. A pod (in the k8s sense) doesn't really have an ordering. It's a group of containers with shared resources (iiuc).
This looks more like k8s init containers. So if you want to cast this into an object structure we'd either need a CephContainerBase and two children CephContainer (with these hooks) and CephContainerHooks (bad name) that excludes these hooks.
Or something like extending CephContainer and then not executing the hooks. That is certainly a bit hacky as far as object inheritance works, but it'd work.

Pragmatically I think this current state would work for now. Not like this is intended to be a generic container runtime.

Comment (sebastian-philipp) [reply src/ceph-daemon/ceph-daemon]: > Hmm I disagree. A pod (in the k8s sense) doesn't really have an ordering. It's a group of containers with shared resources (iiuc).

Ha! Turns out you're right. 

> This looks more like k8s init containers. So if you want to cast this into an object structure we'd either need a CephContainerBase and two children CephContainer (with these hooks) and CephContainerHooks (bad name) that excludes these hooks.
> Or something like extending CephContainer and then not executing the hooks. That is certainly a bit hacky as far as object inheritance works, but it'd work.
> 
> Pragmatically I think this current state would work for now. Not like this is intended to be a generic container runtime.

Hm. Adding CephContainerHooks looks indeed a bit complicated. 

Looks like I had something different in mind. IMO `prestart` and `poststart` for this class still doesn't really look like a good improvement as outlined above, but the resolution might look different: We only make use of `prestart` and `poststart when building the unit files; They're not executed when running the containers. Maybe we need a class that encapsulates these systemd unites? I.e. rename this class to `CephUnit`?

**Comment by sebastian-philipp** [comment]
mypy run-test: commands[0] | mypy cephadm
cephadm:666: error: List item 3 has incompatible type "Optional[str]"; expected "str"
cephadm:683: error: List item 3 has incompatible type "Optional[str]"; expected "str"
cephadm:1648: error: Incompatible types in assignment (expression has type "None", variable has type "str")
Found 3 errors in 1 file (checked 1 source file)

**Pull request merged**